src/azul/chalice.py

```diff
@@ -504,34 +497,120 @@ def swagger_resource(self, file) -> Response:
                                 headers={'Content-Type': content_type},
                                 body=body)
 
-    def threshold(self, *, errors: int, throttles: int):
-        def wrapper(f):
+    # noinspection PyPep8Naming
+    @attr.define(frozen=True)
+    class metric_alarm:
+        """
+        Use this decorator to configure a metric alarm for a Lambda function.
+        This decorator cannot be used to decorate view functions, i.e. functions
+        also decorated with ``@app.route``.
+        """
+        #: The AWS/Lambdas namespace for this alarm
+        metric: LambdaMetric
+        #: The number of failed or throttled lambda invocations that, when
+        # exceeded, will trigger the alarm.
+        threshold: int = attr.ib(default=0)
+        #: The interval (in seconds) at which the alarm threshold is evaluated,
+        # the minimum allowed value is 1 minute, 5 min is the default and the
+        # maximum allowed value is 1 day.
+        period: int = attr.ib(default=5 * 60)
+        lambda_name: Optional[str] = attr.ib(default=None)
+        handler_name: Optional[str] = attr.ib(default=None)
```

Comment:
If I understand correctly, these attributes are only ever set using the `bind` method, and are never set via the constructor. In that case, we could write

```suggestion
        lambda_name: Optional[str] = attr.ib(default=None, init=False)
        handler_name: Optional[str] = attr.ib(default=None, init=False)
```

to make that more obvious.

Reply:
This that's quite work because of the `frozen=True`, since `attr.evolve` returns a new instance of the class, it may not initially define this attributes.